File: 알고리즘 스터디/개인공부/BinarySearch/SteppingStone.py
#https://programmers.co.kr/learn/courses/30/lessons/43236?language=python3

def solution(distance, rocks, n):
    max_distance = distance + 1
    min_distance = 1
    sorted_rocks = [0] + sorted(rocks) + [distance] #시작점 ~ 돌 ~ 도착점
    gap_rocks = [sorted_rocks[i + 1] - sorted_rocks[i] for i in range(len(rocks) + 1)]  #돌 사이의 거리
    answer = distance

    while min_distance < max_distance:
        middle = (max_distance + min_distance) // 2

        # middel내로 거리 만족할 수 있는지 확인
        available_rock = n  #앞으로 더 치울수 있는 돌의 갯수
        c_sum = gap_rocks[0]    #이전 돌까지의 거리
    
        #모든 돌 사이의 거리가 middle보다 크거나 같도록 함
        for gap in gap_rocks[1:]:
            #돌 사이의 거리가 middle보다 작으면 돌을 치워서 middle보다 크게 만들어줌
            if c_sum < middle:
                c_sum += gap
                available_rock -= 1
                if available_rock == -1:
                    break
            else:
                c_sum = gap

        if available_rock == -1:
            max_distance = middle
        else:
            min_distance = middle + 1
            answer = middle

    return answer

if __name__ == '__main__':
    solution(25, [2, 14, 11, 21, 17], 2)

File: 알고리즘 스터디/개인공부/BinarySearch/SteppingStone.java
/* https://school.programmers.co.kr/learn/courses/30/lessons/43236 */

import java.util.Arrays;

public class SteppingStone {
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;
        int start = 1;
        int end = 1000000000;
        Arrays.sort(rocks);

        while (start < end) {
            int middle = (start + end) / 2;
            int prev = 0;
            int remain = n;

            for(int i = 0;i < rocks.length; i++){
                if(rocks[i] - prev < middle){
                    remain--;
                    continue;
                }

                prev = rocks[i];
            }

            // 제일 마지막 돌
            if(distance - prev < middle)
                remain--;   // 제일 마지막 돌 바로 앞에 있는 돌 제거

            if(remain >= 0) {
                start = middle + 1;
                answer = middle > answer?middle:answer;
            }
            else
                end = middle;

        }

        return answer;
    }

    public static void main(String[] args) {
        new SteppingStone().solution(25, new int[]{2, 14, 11, 21, 17}, 2);   // 4
    }
}
